test/extensions/filters/listener/original_dst/original_dst_integration_test.cc

```diff
@@ -0,0 +1,134 @@
+  // Only do this for IPv4 as we can have many 127.0.0.x addresses listening on 0.0.0.0
+  if (version_ != Network::Address::IpVersion::v4) {
+    return;
+  }
+  auto address = Network::Test::getAnyAddress(Network::Address::IpVersion::v4);
+  Network::DownstreamTransportSocketFactoryPtr factory =
+      upstream_tls_ ? createUpstreamTlsContext(upstreamConfig())
+                    : Network::Test::createRawBufferDownstreamSocketFactory();
+  fake_upstreams_.emplace_back(std::make_unique<FakeUpstream>(std::move(factory), address,
+                                                              configWithType(upstreamProtocol())));
+
+  config_helper_.addConfigModifier(
+      setOriginalDstCluster(fake_upstreams_[0]->localAddress()->ip()->port()));
+  initialize();
+
+  const int32_t kMaxConnections = 100;
+  for (int i = 0; i < kMaxConnections; ++i) {
+    Network::ClientConnectionPtr connection(dispatcher_->createClientConnection(
+        Network::Utility::resolveUrl(fmt::format("tcp://127.0.0.{}:{}", i + 1, lookupPort("http"))),
+        Network::Address::InstanceConstSharedPtr(), Network::Test::createRawBufferSocket(), nullptr,
+        nullptr));
+
+    connection->enableHalfClose(enableHalfClose());
+
+    codec_client_ = makeHttpConnection(std::move(connection));
+    auto response =
+        sendRequestAndWaitForResponse(default_request_headers_, 0, default_response_headers_, 0);
+    ASSERT_TRUE(fake_upstream_connection_->close());
+    fake_upstream_connection_.reset();
+    codec_client_->close();
+  }
+}
+
+class OriginalDstTcpProxyIntegrationTest
+    : public testing::TestWithParam<Network::Address::IpVersion>,
+      public BaseIntegrationTest {
+public:
+  OriginalDstTcpProxyIntegrationTest()
+      : BaseIntegrationTest(GetParam(), ConfigHelper::tcpProxyConfig()) {
+    enableHalfClose(true);
+  }
+};
+
+TEST_P(OriginalDstTcpProxyIntegrationTest, TestManyConnections) {
+  if (version_ != Network::Address::IpVersion::v4) {
+    return;
+  }
+
+  // Create an upstream on 0.0.0.0
+  auto address = Network::Test::getAnyAddress(Network::Address::IpVersion::v4);
+  Network::DownstreamTransportSocketFactoryPtr factory =
+      upstream_tls_ ? createUpstreamTlsContext(upstreamConfig())
+                    : Network::Test::createRawBufferDownstreamSocketFactory();
+  fake_upstreams_.emplace_back(std::make_unique<FakeUpstream>(std::move(factory), address,
+                                                              configWithType(upstreamProtocol())));
+
+  config_helper_.addConfigModifier(
+      setOriginalDstCluster(fake_upstreams_[0]->localAddress()->ip()->port()));
+  initialize();
+
+  const int32_t kMaxConnections = 100;
+  for (int i = 0; i < kMaxConnections; ++i) {
+    // Set up the connection
+    IntegrationTcpClientPtr tcp_client = makeTcpConnection(
+        lookupPort("listener_0"), nullptr, nullptr, fmt::format("127.0.0.{}", i + 1));
+    FakeRawConnectionPtr fake_upstream_connection;
```

Comment:
Should we validate the fake upstream connection's local address, also the overrides port?